Split non macro portion of unused_doc_comment from macro part into two passes/lints

yaahc opened the pull request:
## Motivation

This change is motivated by the needs of the [spandoc library](https://github.com/yaahc/spandoc). The specific use case is that my macro is removing doc comments when an attribute is applied to a fn with doc comments, but I would like the lint to still appear when I forget to add the `#[spandoc]` attribute to a fn, so I don't want to have to silence the lint globally.

## Approach

This change splits the `unused _doc_comment` lint into two lints, `unused_macro_doc_comment` and `unused_doc_comment`. The non macro portion is moved into an `early_lint_pass` rather than a pre_expansion_pass. This allows proc macros to silence `unused_doc_comment` warnings by either adding an attribute to silence it or by removing the doc comment before the early_pass runs.

The `unused_macro_doc_comment` lint however will still be impossible for proc-macros to silence, but the only alternative that I can see is to remove this lint entirely, which I don't think is acceptable / is a decision I'm not comfortable making personally, so instead I opted to split the macro portion of the check into a separate lint so that it can be silenced globally with an attribute if necessary without needing to globally silence the `unused_doc_comment` lint as well, which is still desireable.

fixes https://github.com/rust-lang/rust/issues/67838

rust-highfive commented:
r? @ecstatic-morse

(rust_highfive has picked a reviewer for you, use r? to override)

Manishearth commented:
@bors r+

 I was initially concerned that this would break doc comments for macros, but this is just a lint, it's fine. We should not be linting about pretty much _anything_ pre-proc macros except for things which affect lexing, which is true for the other pre-proc-macro pass.

bors commented:
:pushpin: Commit 03863236992e376a87c6d5efb3961be11721d85a has been approved by `Manishearth`

<!-- @bors r=Manishearth 03863236992e376a87c6d5efb3961be11721d85a -->
<!-- homu: {"type":"Approved","sha":"03863236992e376a87c6d5efb3961be11721d85a","approver":"Manishearth"} -->

yaahc commented:
This does break a change made last year to improve diagnostics on macros

```
tree cabd6022a0ec9a157fdce48be26ddc408d626219
parent 018d4d265fb40ea4301da5dc339bff962a6faa57
author Andy Russell <arussell123@gmail.com> Thu Jan 24 15:49:03 2019 -0500
committer Andy Russell <arussell123@gmail.com> Fri Mar 8 12:39:50 2019 -0500

expand unused doc comment diagnostic

Report the diagnostic on macro expansions, and add a label indicating
why the comment is unused.
```

```diff
@@ -802,7 +802,14 @@ impl LintPass for UnusedDocComment {
 }
 
 impl UnusedDocComment {
-    fn warn_if_doc(&self, cx: &EarlyContext, attrs: &[ast::Attribute]) {
+    fn warn_if_doc(
+        &self,
+        cx: &EarlyContext<'_>,
+        node_span: Span,
+        node_kind: &str,
+        is_macro_expansion: bool,
+        attrs: &[ast::Attribute]
+    ) {
         let mut attrs = attrs.into_iter().peekable();
 
         // Accumulate a single span for sugared doc comments.
@@ -825,27 +832,51 @@ impl UnusedDocComment {
             let span = sugared_span.take().unwrap_or_else(|| attr.span);
 
             if attr.name() == "doc" {
-                cx.struct_span_lint(
-                    UNUSED_DOC_COMMENTS,
-                    span,
-                    "doc comment not used by rustdoc",
-                ).emit();
+                let mut err = cx.struct_span_lint(UNUSED_DOC_COMMENTS, span, "unused doc comment");
+
+                err.span_label(
+                    node_span,
+                    format!("rustdoc does not generate documentation for {}", node_kind)
+                );
+
+                if is_macro_expansion {
+                    err.help("to document an item produced by a macro, \
+                              the macro must produce the documentation as part of its expansion");
+                }
+
+                err.emit();
             }
         }
     }
 }
```

I'd like to not break this if possible, is it possible for a lint to exist in two passes, I really only need the lint on items in fns to be done after expansion so I can strip out doc comments in fn bodies.

Manishearth commented:
@yaahc write two passes instead, one that specifically looks for doc comments on macro expansions only. multiple passes are allowed to produce the same lint

@bors r-

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=20778&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20778/logs/70), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20778/logs/70)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-12T03:42:50.7331339Z ========================== Starting Command Output ===========================
2020-02-12T03:42:50.7333350Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/c0481a41-7933-4ed4-8154-56d13d486a72.sh
2020-02-12T03:42:50.7333389Z 
2020-02-12T03:42:50.7336886Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-12T03:42:50.7342746Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-12T03:42:50.7344208Z Task         : Get sources
2020-02-12T03:42:50.7344254Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-12T03:42:50.7344287Z Version      : 1.0.0
2020-02-12T03:42:50.7344540Z Author       : Microsoft
---
2020-02-12T03:42:51.8331524Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-12T03:42:51.8345576Z ##[command]git config gc.auto 0
2020-02-12T03:42:51.8348006Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-12T03:42:51.8351250Z ##[command]git config --get-all http.proxy
2020-02-12T03:42:51.8358222Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69084/merge:refs/remotes/pull/69084/merge
---
2020-02-12T04:12:51.4997231Z Assembling stage1 compiler (x86_64-unknown-linux-gnu)
2020-02-12T04:12:51.5016978Z Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2020-02-12T04:12:51.9532691Z    Compiling cc v1.0.50
2020-02-12T04:12:51.9533062Z    Compiling core v0.0.0 (/checkout/src/libcore)
2020-02-12T04:12:52.3733717Z thread 'rustc' panicked at 'src/librustc_lint/context.rs:179: duplicate specification of lint unused_doc_comments', src/librustc/util/bug.rs:37:26
2020-02-12T04:12:52.3741021Z 
2020-02-12T04:12:52.3745001Z error: internal compiler error: unexpected panic
2020-02-12T04:12:52.3747773Z 
2020-02-12T04:12:52.3751935Z note: the compiler unexpectedly panicked. this is a bug.
2020-02-12T04:12:52.3751935Z note: the compiler unexpectedly panicked. this is a bug.
2020-02-12T04:12:52.3753299Z 
2020-02-12T04:12:52.3759067Z note: we would appreciate a bug report: ***/blob/master/CONTRIBUTING.md#bug-reports
2020-02-12T04:12:52.3763368Z note: rustc 1.43.0-nightly (932f50894 2020-02-12) running on x86_64-unknown-linux-gnu
2020-02-12T04:12:52.3764497Z 
2020-02-12T04:12:52.3764497Z 
2020-02-12T04:12:52.3770128Z note: compiler flags: -Z macro-backtrace -Z binary-dep-depinfo -Z force-unstable-if-unmarked -C opt-level=3 -C codegen-units=1 -C debuginfo=0 -C link-args=-Wl,-rpath,$ORIGIN/../lib -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -C debug-assertions=n --crate-type lib
2020-02-12T04:12:52.3774086Z note: some of the compiler flags provided by cargo are hidden
2020-02-12T04:12:52.3815228Z 
2020-02-12T04:12:52.4168956Z error: could not compile `core`.
2020-02-12T04:12:52.4179956Z warning: build failed, waiting for other jobs to finish...
---
2020-02-12T04:12:58.7923765Z   local time: Wed Feb 12 04:12:58 UTC 2020
2020-02-12T04:12:58.8851931Z   network time: Wed, 12 Feb 2020 04:12:58 GMT
2020-02-12T04:12:58.8854570Z == end clock drift check ==
2020-02-12T04:13:00.3423539Z 
2020-02-12T04:13:00.3518173Z ##[error]Bash exited with code '1'.
2020-02-12T04:13:00.3529979Z ##[section]Finishing: Run build
2020-02-12T04:13:00.3551814Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-12T04:13:00.3553398Z Task         : Get sources
2020-02-12T04:13:00.3553440Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-12T04:13:00.3553498Z Version      : 1.0.0
2020-02-12T04:13:00.3553533Z Author       : Microsoft
2020-02-12T04:13:00.3553533Z Author       : Microsoft
2020-02-12T04:13:00.3553575Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-02-12T04:13:00.3553648Z ==============================================================================
2020-02-12T04:13:00.7926765Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-02-12T04:13:00.7972887Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-12T04:13:00.8089911Z Cleaning up task key
2020-02-12T04:13:00.8090814Z Start cleaning up orphan processes.
2020-02-12T04:13:00.8235017Z Terminate orphan process: pid (3460) (python)
2020-02-12T04:13:00.8878082Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Manishearth commented:
@bors r+

bors commented:
:pushpin: Commit a89a791fd3f3665fc9a63f8b837fafe888d35e69 has been approved by `Manishearth`

<!-- @bors r=Manishearth a89a791fd3f3665fc9a63f8b837fafe888d35e69 -->
<!-- homu: {"type":"Approved","sha":"a89a791fd3f3665fc9a63f8b837fafe888d35e69","approver":"Manishearth"} -->

yaahc commented:
Actually, I think I need to disable the macro prepass lint for `check_stmt`. I think this will be okay because we probably dont have people commonly trying to declare items in macros as statements and document them at the same time...

Manishearth commented:
@bors r- delegate+

bors commented:
:v: @yaahc can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"Manishearth","delegate":"yaahc"} -->

yaahc commented:
So for my needs I would definitely prefer that doc comments on macro calls as statements are still handled in the early_pass so they can be silenced. I tried removing the stmt check in the pre_expansion pass lint with the hope that `expr` / `stmt` checks would catch the attributes on the post expansion version but the attributes seem to disappear and we no longer produce a lint on the macro call within the fn body, even if I changed the macro call to actually expand to an expr.

Could I get some help figuring out how to correctly lint for doc comments on macros as expressions from an early pass lint or decide which would be preferable between not linting in this case at all or linting on this but reporting the lint as `unused_macro_doc_comment` so that it can be silenced independently, either of which would be okay though not ideal solutions for my problem.

yaahc commented:
I went ahead and implemented the version that exports the macro portion of the lint as a separately named lint, in addition to having it implemented as a separate lint pass internally. I figured this would be the most likely solution we'd end up needing.

If someone knows of a way to still appropriately lint on the attributes on macros after they've been expanded then I'm down to put in the work to swap the solution to that, but it looks like prior to this being moved to be a preexpansion pass lint the lint just straight up didn't trigger for macros, so my expectation is that attributes on macros are discarded after expansion, and that this is the only way to preserve the lint but while not requiring a blanket allow on all unused doc comments to silence the macro lint warnings.

Manishearth commented:
@bors r+

bors commented:
:pushpin: Commit df6eb29fa0d4003bd0ec2bbf7ef55e4cfb9e5d5c has been approved by `Manishearth`

<!-- @bors r=Manishearth df6eb29fa0d4003bd0ec2bbf7ef55e4cfb9e5d5c -->
<!-- homu: {"type":"Approved","sha":"df6eb29fa0d4003bd0ec2bbf7ef55e4cfb9e5d5c","approver":"Manishearth"} -->

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=20789&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20789/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20789/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-12T06:45:21.6918066Z ========================== Starting Command Output ===========================
2020-02-12T06:45:21.7189109Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/5183fbc9-d6b9-4868-b3ea-de2f5e170ca1.sh
2020-02-12T06:45:21.7189311Z 
2020-02-12T06:45:21.7192499Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-12T06:45:21.7198998Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-12T06:45:21.7201454Z Task         : Get sources
2020-02-12T06:45:21.7201491Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-12T06:45:21.7201575Z Version      : 1.0.0
2020-02-12T06:45:21.7201612Z Author       : Microsoft
---
2020-02-12T06:45:22.5581913Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-12T06:45:22.5709064Z ##[command]git config gc.auto 0
2020-02-12T06:45:22.5772649Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-12T06:45:22.5835238Z ##[command]git config --get-all http.proxy
2020-02-12T06:45:22.5975697Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69084/merge:refs/remotes/pull/69084/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

petrochenkov commented:
@bors rollup-
(I'd like to review this.)

Dylan-DPC commented:
@bors r-

yaahc commented:
Let me know if you need any explanation or there's anything else I can do to help the review along @petrochenkov

Manishearth commented:
Also, for clarity I and @yaahc were discussing this PR out of band. A couple things I should repeat:

Firstly, I think the pre-proc-macro lints should be minimal, ideally only the ones that deal with lex stuff, which isn't true for this lint. The macro part does make sense to run there, however.

Also, I think the case of doc commenting a macro invocation within a statement is rare enough that it's fine to skip, given that the other lint will catch it if the macro expands to a statement. Items within functions don't get documented _anyway_.

(It actually might be worth expanding the lint to cover any items defined in a function, though it's tricky since impl and trait items still get documented)

yaahc commented:
> Also, I think the case of doc commenting a macro invocation within a statement is rare enough that it's fine to skip, given that the other lint will catch it if the macro expands to a statement. Items within functions don't get documented anyway.

I tried testing this and it doesn't actually seem to be the case, it seems like attributes on macros are discarded after expansion rather than being applied to the first item within the macro. This is why I added back the macro check on `stmts` in the pre_pass and split it into another macro, so we still get the warning but it can be silenced independently.

petrochenkov commented:
I certainly agree that `unused_doc_comments` shouldn't be a pre-expansion pass.
I however don't think `unused_macro_doc_comments` should exist.

As it was said previously, the only case that's not reported if `unused_macro_doc_comments` is removed is something like this
```rust
/// Doc comment.
macro_call!();
```
, and all cases like this can caught by reporting the same `unused_doc_comments` lint from `fn check_attributes` in `librustc_expand/expand.rs` which is called for all attributes disappearing in macro expansion.
This can be done with `fn buffer_lint` if I'm not mistaken (the lint infra is pretty volatile and was changed multiple times recently).

yaahc commented:
> As it was said previously, the only case that's not reported if `unused_macro_doc_comments` is removed is something like this

@petrochenkov I don't think this is true, like I mentioned earlier experimental tests, the checks on stmts and expressions do not catch unused doc comments after expansion. It seems like the attributes don't get applied to the items within the macro upon expansion, which is likely the main reason that you cannot document items generated by a macro with a comment outside the macro, which is why they added the extra diagnostic.

For example, the following would not trigger a lint if we remove the macro stuff

```rust
/// print some text
println!("doing a thing");
```

I'm okay with removing it either way, because the unused_macro_doc_comment doesn't do anything for me, but we would be removing the extra diagnostics that `Andy Russell <arussell123@gmail.com>` aka @euclio felt the need to add

petrochenkov commented:
>the checks on stmts and expressions do not catch unused doc comments after expansion.

That's true, but `fn check_attributes` in `librustc_expand/expand.rs` happens *during* expansion, not after expansion. So it allows to check attributes that no longer exist after expansion.

Some other checks like surface attribute well-formedness are also performed in that place:
```rust
macro_rules! empty { () => () }

fn main() {
    // The error below is reported despite the inline attribute disappearing during expansion.
    #[inline = "nonsense"] // error: attribute must be of the form `#[inline]` or `#[inline(always|never)]`
    empty!();
}
```

yaahc commented:
> That's true, but fn check_attributes in librustc_expand/expand.rs happens during expansion, not after expansion. So it allows to check attributes that no longer exist after expansion.

I'm not sure what to make of this. I poked around in those fns a little and it looks like the check for attribute form is skipped for doc comments

```rust
    if attr.is_doc_comment() {
        return;
    }
```

in the `check_meta` fn called from `check_attributes`. I'm not sure what change you're proposing at this point.

petrochenkov commented:
@yaahc 
>So I'm not sure what change you're proposing at this point

Something like [this](https://github.com/petrochenkov/rust/commits/unudoc), but more careful.

>I don't understand how you can tell from check_attributes if you're an attribute of an item that still exists or not.

`check_attributes` in `expand.rs` is *never* an item that still exists.
(Except for the case in `visit_attribute` where people writing it had little idea of what they were doing.)

euclio commented:
For context, I originally added the warnings because this is a somewhat common mistake:

```rust
macro_rules! foo {
    ($( $tokens:tt )*) => { $( $tokens )* }
}

/// Docs for struct S
foo! {
    pub struct S;
}
```

If we can achieve the same warnings without making the lint an early pass, great!

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21506&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21506/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21506/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-21T22:45:13.8193451Z ========================== Starting Command Output ===========================
2020-02-21T22:45:13.8196547Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/c27abfde-37d6-4761-8c36-bd0ae759c838.sh
2020-02-21T22:45:13.8197192Z 
2020-02-21T22:45:13.8202121Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-21T22:45:13.8222508Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-21T22:45:13.8225466Z Task         : Get sources
2020-02-21T22:45:13.8282809Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-21T22:45:13.8283263Z Version      : 1.0.0
2020-02-21T22:45:13.8283427Z Author       : Microsoft
---
2020-02-21T22:45:14.8308447Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-21T22:45:14.8317343Z ##[command]git config gc.auto 0
2020-02-21T22:45:14.8324291Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-21T22:45:14.8330862Z ##[command]git config --get-all http.proxy
2020-02-21T22:45:14.8339017Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69084/merge:refs/remotes/pull/69084/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

yaahc commented on src/librustc_expand/expand.rs:
```diff
@@ -1090,6 +1092,15 @@ impl<'a, 'b> InvocationCollector<'a, 'b> {
+                self.cx.parse_sess.buffer_lint_with_diagnostic(
+                    &UNUSED_DOC_COMMENTS,
+                    attr.span,
+                    ast::CRATE_NODE_ID,
```
Is there any way to get a more specific node_id here or has this info been lost by this point? @eddyb raised the point that this will only honor crate wide attribute settings. I don't mind doing extra work to get a more specific ID here but if it can't be helped then that's fine.

petrochenkov replied:
It's ok to use `CRATE_NODE_ID` here.
It's probably possible to make more specific node ids available here, but the necessary infra will be large than this whole PR.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21509&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21509/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21509/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-22T00:02:34.9880461Z ========================== Starting Command Output ===========================
2020-02-22T00:02:34.9883967Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/b4761e97-69a6-4c6d-a967-6fff8261363d.sh
2020-02-22T00:02:34.9884391Z 
2020-02-22T00:02:34.9888154Z ##[section]Finishing: Disable git automatic line ending conversion
2020-02-22T00:02:34.9908133Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/69084/merge to s
2020-02-22T00:02:34.9911593Z Task         : Get sources
2020-02-22T00:02:34.9911922Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-02-22T00:02:34.9912238Z Version      : 1.0.0
2020-02-22T00:02:34.9912471Z Author       : Microsoft
---
2020-02-22T00:02:36.1817737Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-02-22T00:02:36.1826065Z ##[command]git config gc.auto 0
2020-02-22T00:02:36.1830856Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-02-22T00:02:36.1836201Z ##[command]git config --get-all http.proxy
2020-02-22T00:02:36.1845465Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/69084/merge:refs/remotes/pull/69084/merge
---
2020-02-22T00:06:01.0494631Z Successfully built 34a69d9dd0ee
2020-02-22T00:06:01.0587213Z Successfully tagged rust-ci:latest
2020-02-22T00:06:01.1578721Z Built container sha256:34a69d9dd0ee8690fc6db18ad26b8d3cd2756daa6b573f75d41e978b1f5a223a
2020-02-22T00:06:01.1597820Z Uploading finished image to https://rust-lang-ci-sccache2.s3.amazonaws.com/docker/d42805294727747964a78ce70c906cffcf151e0ac7404a3765966e44959ef24b29c11dc98d21205141b0951a747cfe58c33808b212e0a5b1f239c2c509aa3b06
2020-02-22T00:06:47.3558149Z upload failed: - to s3://rust-lang-ci-sccache2/docker/d42805294727747964a78ce70c906cffcf151e0ac7404a3765966e44959ef24b29c11dc98d21205141b0951a747cfe58c33808b212e0a5b1f239c2c509aa3b06 An error occurred (InvalidAccessKeyId) when calling the CreateMultipartUpload operation: The AWS Access Key Id you provided does not exist in our records.
2020-02-22T00:06:47.8290511Z [CI_JOB_NAME=x86_64-gnu-llvm-7]
2020-02-22T00:06:47.8330180Z == clock drift check ==
2020-02-22T00:06:47.8343656Z   local time: Sat Feb 22 00:06:47 UTC 2020
2020-02-22T00:06:48.1291676Z   network time: Sat, 22 Feb 2020 00:06:48 GMT
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

petrochenkov commented on src/librustc_lint/builtin.rs:
```diff
@@ -738,21 +738,22 @@ impl EarlyLintPass for DeprecatedAttr {
-    Warn,
-    "detects doc comments that aren't used by rustdoc"
+trait UnusedDocCommentExt {
+    fn warn_if_doc(
```
The trait is unnecessary, I only used it to minimize the diff.
This can be a free function.

petrochenkov commented:
r=me with the trait removed.

yaahc commented:
r=me

yaahc commented:
@bors r=me

yaahc commented:
@bors r+

bors commented:
:pushpin: Commit 494dd0b71936813e3a5565d5ea88c6fdf07240e3 has been approved by `yaahc`

<!-- @bors r=yaahc 494dd0b71936813e3a5565d5ea88c6fdf07240e3 -->
<!-- homu: {"type":"Approved","sha":"494dd0b71936813e3a5565d5ea88c6fdf07240e3","approver":"yaahc"} -->

bors commented:
:hourglass: Testing commit 494dd0b71936813e3a5565d5ea88c6fdf07240e3 with merge 3cca84e4e6b1fe2d8b85dc819bb0047080e19c73...
<!-- homu: {"type":"BuildStarted","head_sha":"494dd0b71936813e3a5565d5ea88c6fdf07240e3","merge_sha":"3cca84e4e6b1fe2d8b85dc819bb0047080e19c73"} -->

rust-highfive commented:
The job `dist-various-2` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21574&view=logs&jobId=910a3b72-54f3-5d99-d01c-16c9c22f04ca) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21574/logs/1382), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21574/logs/1382)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-22T22:42:01.4127771Z [RUSTC-TIMING] hashbrown test:false 0.595
2020-02-22T22:42:01.5698486Z warning: dropping unsupported crate type `dylib` for target `wasm32-wasi`
2020-02-22T22:42:01.5698803Z 
2020-02-22T22:42:01.8724606Z error: unused doc comment
2020-02-22T22:42:01.8725911Z   --> src/libstd/sys/wasi/fd.rs:16:5
2020-02-22T22:42:01.8726888Z    |
2020-02-22T22:42:01.8728122Z 16 |       /// SAFETY: `IoSliceMut` and `IoVec` have exactly the same memory layout
2020-02-22T22:42:01.8729876Z 17 | /     unsafe {
2020-02-22T22:42:01.8730645Z 18 | |         mem::transmute(a)
2020-02-22T22:42:01.8731343Z 19 | |     }
2020-02-22T22:42:01.8732062Z    | |_____- rustdoc does not generate documentation for expressions
2020-02-22T22:42:01.8732062Z    | |_____- rustdoc does not generate documentation for expressions
2020-02-22T22:42:01.8732588Z    |
2020-02-22T22:42:01.8733123Z    = note: `-D unused-doc-comments` implied by `-D warnings`
2020-02-22T22:42:01.8733608Z 
2020-02-22T22:42:01.8739246Z error: unused doc comment
2020-02-22T22:42:01.8739836Z   --> src/libstd/sys/wasi/fd.rs:25:5
2020-02-22T22:42:01.8740251Z    |
2020-02-22T22:42:01.8741161Z 25 |       /// SAFETY: `IoSlice` and `CIoVec` have exactly the same memory layout
2020-02-22T22:42:01.8742871Z 26 | /     unsafe {
2020-02-22T22:42:01.8743768Z 27 | |         mem::transmute(a)
2020-02-22T22:42:01.8744441Z 28 | |     }
2020-02-22T22:42:01.8745355Z    | |_____- rustdoc does not generate documentation for expressions
---
2020-02-22T22:42:03.4639687Z   local time: Sat Feb 22 22:42:03 UTC 2020
2020-02-22T22:42:03.6027931Z   network time: Sat, 22 Feb 2020 22:42:03 GMT
2020-02-22T22:42:03.6032108Z == end clock drift check ==
2020-02-22T22:42:05.6200724Z 
2020-02-22T22:42:05.6271599Z ##[error]Bash exited with code '1'.
2020-02-22T22:42:05.6334040Z ##[section]Starting: Checkout rust-lang/rust@auto to s
2020-02-22T22:42:05.6338116Z ==============================================================================
2020-02-22T22:42:05.6338430Z Task         : Get sources
2020-02-22T22:42:05.6338749Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21574)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21574","builder_name":"checks-azure"} -->

yaahc commented on src/libstd/sys/wasi/fd.rs:
```diff
@@ -13,7 +13,7 @@ pub struct WasiFd {
     assert_eq!(mem::size_of::<IoSliceMut<'_>>(), mem::size_of::<wasi::Iovec>());
     assert_eq!(mem::align_of::<IoSliceMut<'_>>(), mem::align_of::<wasi::Iovec>());
-    /// SAFETY: `IoSliceMut` and `IoVec` have exactly the same memory layout
+    // SAFETY: `IoSliceMut` and `IoVec` have exactly the same memory layout
```
This caused a build failure due to the warning, I don't quite understand how this hasn't been caught in the past though.

yaahc commented:
@bors r+

bors commented:
:pushpin: Commit 40c67221e272c007cfae407d4398b166754491c9 has been approved by `yaahc`

<!-- @bors r=yaahc 40c67221e272c007cfae407d4398b166754491c9 -->
<!-- homu: {"type":"Approved","sha":"40c67221e272c007cfae407d4398b166754491c9","approver":"yaahc"} -->

yaahc commented:
@bors r+

bors commented:
:pushpin: Commit 09bc5e3d969a154ffcbeb6827a901d36a6a854eb has been approved by `yaahc`

<!-- @bors r=yaahc 09bc5e3d969a154ffcbeb6827a901d36a6a854eb -->
<!-- homu: {"type":"Approved","sha":"09bc5e3d969a154ffcbeb6827a901d36a6a854eb","approver":"yaahc"} -->

bors commented:
:hourglass: Testing commit 09bc5e3d969a154ffcbeb6827a901d36a6a854eb with merge ca07a2365ddd2a24a92a9ed4a02ec78f9d82deeb...
<!-- homu: {"type":"BuildStarted","head_sha":"09bc5e3d969a154ffcbeb6827a901d36a6a854eb","merge_sha":"ca07a2365ddd2a24a92a9ed4a02ec78f9d82deeb"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21586)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21586","builder_name":"checks-azure"} -->

petrochenkov commented:
@yaahc 
Meta: the convention is to `@bors r=petrochenkov` on behalf of the reviewer instead of `@bors r+`ing on behalf of yourself.

The CI failure looks spurious.
@bors r+

petrochenkov commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #69218

bors commented:
:pushpin: Commit 09bc5e3d969a154ffcbeb6827a901d36a6a854eb has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 09bc5e3d969a154ffcbeb6827a901d36a6a854eb -->
<!-- homu: {"type":"Approved","sha":"09bc5e3d969a154ffcbeb6827a901d36a6a854eb","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 09bc5e3d969a154ffcbeb6827a901d36a6a854eb with merge 0cbf1a602ead3e16296c5a43ab78ec5044e309f3...
<!-- homu: {"type":"BuildStarted","head_sha":"09bc5e3d969a154ffcbeb6827a901d36a6a854eb","merge_sha":"0cbf1a602ead3e16296c5a43ab78ec5044e309f3"} -->

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=21621&view=logs&jobId=10c53b70-9ae7-511a-b369-2febdb4c7298) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21621/logs/511), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/21621/logs/511)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21621)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21621","builder_name":"checks-azure"} -->

petrochenkov commented:
@bors retry

bors commented:
:hourglass: Testing commit 09bc5e3d969a154ffcbeb6827a901d36a6a854eb with merge b1f395de642e8be7bcbbd2bd8aaadab715851f49...
<!-- homu: {"type":"BuildStarted","head_sha":"09bc5e3d969a154ffcbeb6827a901d36a6a854eb","merge_sha":"b1f395de642e8be7bcbbd2bd8aaadab715851f49"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21630)
Approved by: petrochenkov
Pushing b1f395de642e8be7bcbbd2bd8aaadab715851f49 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=21630"},"merge_sha":"b1f395de642e8be7bcbbd2bd8aaadab715851f49"} -->
